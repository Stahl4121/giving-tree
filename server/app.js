// Stripe Server

const app = require('express')();
const bodyParser = require('body-parser');
const stripe = require('stripe')('sk_test_avCwuzIvg45JxkjItTyqRGH600JEoryzyP');
const admin = require('firebase-admin');
const serviceAccount = require("./secret/givingtree-cfs-firebase-adminsdk-grrpd-764871e22c.json");

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount),
  databaseURL: "https://givingtree-cfs.firebaseio.com"
});

app.use(bodyParser.text());
app.use(bodyParser.urlencoded({ extended: true }));

app.post('/charge', async (req, res) => {
  let source = req.body.split(' amount: ')[0];
  let amount = req.body.split(' amount: ')[1].split(' description: ')[0];
  let description = req.body.split(' amount: ')[1].split(' description: ')[1].split(' account: ')[0];
  let account = req.body.split(' amount: ')[1].split(' description: ')[1].split(' account: ')[1];
  try {
    let { status } = await stripe.charges.create({
      amount: amount,
      currency: 'usd',
      description: description,
      source: source,
    }, {
      stripeAccount: account,
    }).then(function (charge) {
      // asynchronously called
      return charge;
    });

    res.json({ status });
  } catch (err) {
    console.log(err);
    res.status(500).end();
  }
});

app.post('/create', async (req, res) => {
  try {
    let connected_account_id = await stripe.oauth.token({
      grant_type: 'authorization_code',
      code: req.body,
    }).then(function (response) {
      // asynchronously called
      return response.stripe_user_id;
    });
    return res.json({ 'stripeId': connected_account_id });

  } catch (err) {
    res.status(500).end;
  }
})



//Get current user's ID token
//If metaadmin, call "getUserByEmail" w/ the community foundation email, update status to accepted/rejected
//If not metaadmin, get current user's ID token and set custom claim 'cf' to true
app.post('/setCustomClaims', (req, res) => {

  // Get the ID token passed.
  const idToken = req.body.idToken;
  const cfEmail = req.body.cfEmail;
  const cfStatus = req.body.cfStatus;

  // Verify the ID token and decode its payload.
  admin.auth().verifyIdToken(idToken).then((claims) => {
    console.log('verified token\n');

    //Check if metaadmin
    if (typeof claims.admin !== 'undefined' && claims.admin) {
      //Check that cfEmail and cfStatus were provided in request
      if (typeof cfEmail !== 'undefined' && typeof cfStatus !== 'undefined'
        && cfEmail !== '' && cfStatus !== '' && typeof claims.cfId !== 'undefined' ) {
        //Get the CF User and update their permissions
        admin.auth().getUserByEmail(cfEmail).then((user) => {
          if (cfStatus === 'current') {
            admin.auth().setCustomUserClaims(user.uid, { cfId: claims.cfId, status: 'current' })
              .then(function () {
                // Tell client to refresh token on user.
                res.end(JSON.stringify({ status: 'success' }));
              })
              .catch((error) => {
                res.end(JSON.stringify({ status: error }));
              });
          }
          else if (cfStatus === 'denied') {
              admin.auth().setCustomUserClaims(user.uid, { cfId: claims.cfId, status: 'denied' })
                .then(function () {
                  // Tell client to refresh token on user.
                  res.end(JSON.stringify({ status: 'success' }));
                })
                .catch((error) => {
                  res.end(JSON.stringify({ status: error }));
                });
          }
        })
          .catch((error) => {
            res.end(JSON.stringify({ status: error }));
          });
      }
    }
    else {
      console.log('not metaadmin\n');
      //Not metaadmin, so we this post must be made from CF Account Request
      //Get and store the cf document record id in the claims object
      admin.auth().getUser(claims.sub)
        .then((user) => {
          console.log('got user\n');
          admin.firestore().collection('communityFoundations').where('personal_email', '==', user.email)
            .get()
            .then(function (querySnapshot) {
              querySnapshot.forEach(function (doc) {
              console.log('got cf!\n' );
                admin.auth().setCustomUserClaims(claims.sub, { cfId: doc.id, status: 'requested' })
                  .then(() => {
                  console.log('set cc!\n');
                    // Tell client to refresh token on user.
                    res.end(JSON.stringify({ status: 'success' }));
                  })
                  .catch((error) => {
                  console.log('failed cc!\n');

                    res.end(JSON.stringify({ status: error }));
                  });
                  console.log('after cc\n');
              });
            })
            .catch((error) => {
              res.end(JSON.stringify({ status: error }));
            });
        })
        .catch((error) => {
          res.end(JSON.stringify({ status: error }));
        });
        console.log('end of not metaadmin\n');
    }
  })
  .catch((error) => {
    console.log('NOOO at the end\n');
    res.end(JSON.stringify({ status: error }));
  });
});

app.listen(process.env.PORT);
// app.listen(9000, () => console.log('Listening on port 9000')); 	