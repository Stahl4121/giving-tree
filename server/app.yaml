#TODO: Figure out what else should go in here
runtime: nodejs10
