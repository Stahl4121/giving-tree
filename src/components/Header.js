import React from 'react';
import Search from './Search';

export default function Header() {
  return (
    <div>
      <p>Giving Tree Header!</p>
      <Search/>
    </div>
  );
}
