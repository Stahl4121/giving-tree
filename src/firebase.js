//Reasoning for this file: https://stackoverflow.com/questions/48492047/where-do-i-initialize-firebase-app-in-react-application

// Firebase App (the core Firebase SDK) is always required and
// must be listed before other Firebase SDKs
import * as firebase from "firebase/app";

// Add the Firebase services that you want to use
import "firebase/auth";
import "firebase/firestore";

//Import everything, not recommended for production apps
//import firebase from "firebase";

const firebaseConfig = {
  apiKey: "AIzaSyAkeCK6w-qbvHzaOdEyg_PtpGQbElAv0_M",
  authDomain: "givingtree-cfs.firebaseapp.com",
  databaseURL: "https://givingtree-cfs.firebaseio.com",
  projectId: "givingtree-cfs",
  storageBucket: "givingtree-cfs.appspot.com",
  messagingSenderId: "709659064913",
  appId: "1:709659064913:web:d9decf22beb442f2e79670",
  measurementId: "G-KZ6PNG9K64"
};


firebase.initializeApp(firebaseConfig);

export default !firebase.apps.length ? firebase.initializeApp(firebaseConfig) : firebase.app();