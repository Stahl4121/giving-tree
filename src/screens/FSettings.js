import React, { useEffect } from 'react';
import AuthUserContext from '../auth/context.js';

import {firestore as FIRESTORE} from "firebase/app";
import firebase from '../firebase.js';

import EditableData from '../components/FSettingsListEditable.js'
import NonEditableData from '../components/FSettingList.js'


export default function FSettings() {
 
  // Set tab title
  useEffect(() => { document.title = 'Settings- Giving Tree'; }, []);
  var db = firebase.firestore();

  const [isEdit, setEdit] = React.useState(false);
  const user = React.useContext(AuthUserContext);

  const [getData, setGetData] = React.useState(true);

  // Foundation info
  const [cfInfo, setCfInfo] = React.useState({
    name: '',
    public_email: '',
    public_phone: '',
    foundation_url: '',
    fname_contact: '',
    lname_contact: '',
    personal_email: '',
    personal_phone: '',
    status: '',
    date_requested: '',
    date_approved: '',
    date_denied: '',
    date_deactivated: '',
  });

  
  function toggleEdit()
  {
    if(isEdit){
      setEdit(false);
    }else{
      setEdit(true);
    }
  }

  React.useEffect(() => {
    if(getData){
      if((user !== null)){
        FSettingsMethods().loadData(user, db, (doc) => {
          setCfInfo(doc.data());
          setGetData(false);
        });
      }
    }
  }, [getData, user, db]);
  
  if(isEdit){
    return(
      <EditableData
        isEdit={isEdit}
        cfInfo={cfInfo}
        toggleEdit={toggleEdit}
        onSubmit={(changedText) => FSettingsMethods().onSubmit(cfInfo, changedText, user, db, () => {
          setGetData(true);
          setEdit(false);
        })}
        functionLoadData={() => FSettingsMethods().loadData}
      />
    );
  }else{
    return(
      <NonEditableData
        cfInfo={cfInfo}
        toggleEdit={toggleEdit}
        onSubmit={(changedText) => FSettingsMethods().onSubmit(cfInfo, changedText, user, db, () => {
          setGetData(true, () => setEdit(false));
        })}
        toggleAccountActive={() => FSettingsMethods().toggleAccountActive(db, user, cfInfo, () => {
          setGetData(true);
        })}
      />
    );
  }
} 


// All this crapola is outside the original functional component for testing purposes.
// They say you should write your tests for your code, but since when do "they" know what they're talking about? *sobs*
export const FSettingsMethods = () => {
  const loadData = (user, db, callback) =>{
    // Foundation query
    db.collection('communityFoundations').where('personal_email','==', user.email)
      .get()
      .then(function(querySnapshot) {
        querySnapshot.forEach(function(doc) {
          // doc.data() is never undefined for query doc snapshots
          //console.log(doc.id, " => ", doc.data());
          callback(doc);
        });
      })
      .catch(function(error) {
        console.log("Error getting documents: ", error);
      });
  }
  
  
  const onSubmit = (cfInfo, changedText, user, db, callback) =>{
    var temp = {...cfInfo};
    for(const key in changedText){
      temp = { ...temp, [key]: changedText[key] };
    }
  
    db.collection('communityFoundations').where('personal_email','==',user.email)
    .get()
    .then(function(querySnapshot) {
      querySnapshot.forEach(function(doc) {
        //console.log(doc.id, " => ", doc.data());
        // Build doc ref from doc.id
        db.collection("communityFoundations").doc(doc.id).update({
          name: temp.name,
          public_email: temp.public_email,
          public_phone: temp.public_phone,
          foundation_url: temp.foundation_url,
          fname_contact: temp.fname_contact,
          lname_contact: temp.lname_contact,
          personal_email: temp.personal_email,
          personal_phone: temp.personal_phone
        }).then(
          callback(temp)
        );
        console.log("Document successfully written!");
      });
    })
    .catch(function(error) {
      console.error("Error writing document: ", error);
    });
  }
  
  const toggleAccountActive = (db, user, cfInfo, callback) => {
    if(user){
      db.collection('communityFoundations').where('personal_email','==',user.email)
      .get()
      .then(function(querySnapshot) {
        querySnapshot.forEach(function(doc) {
          const time = (cfInfo.date_deactivated === '') ? FIRESTORE.FieldValue.serverTimestamp() : '';
          db.collection("communityFoundations").doc(doc.id).update(
            {date_deactivated: time},
          ).then(
            () => callback(time)
          );
          console.log("Deactivation successfully toggled!")
        });
      })
      .catch(function(error) {
        console.error("Error writing document: ", error);
      });
    }
  }

  return {
    loadData,
    onSubmit,
    toggleAccountActive,
  };
}


