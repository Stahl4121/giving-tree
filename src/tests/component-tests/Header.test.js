import React from 'react';
import { shallow, mount } from 'enzyme';
import AuthUserContext from '../../auth/context.js';

import FSettings from '../../screens/FSettings.js';
import {firestore as FIRESTORE} from "firebase/app";
import { MemoryRouter } from 'react-router-dom';
import { Header } from '../../components/Header.js';

import { createMemoryHistory } from 'history'
import { jssPreset } from '@material-ui/core';


const { mockFirebase } = require('firestore-jest-mock');
const { mockCollection, mockWhere } = require('firestore-jest-mock/mocks/firestore');



describe('Header logged out', () => {
  let wrap;

  beforeEach(() => {
    wrap = mount(
      <MemoryRouter >
        <Header />
      </MemoryRouter> 
    )
  });

  it('Renders correctly', () => {  
    expect(wrap).toMatchSnapshot();
  });

  it('Opens menu', () => {  
    expect(wrap.find("WithStyles(ForwardRef(Menu))").at(0).props().open).toBeFalsy();
    wrap.find("WithStyles(ForwardRef(IconButton))").at(1).simulate('click');
    expect(wrap.find("WithStyles(ForwardRef(Menu))").at(0).props().open).toBeTruthy();
  });

  it('Closes menu', () => { 
    const history = {
      push : jest.fn()
    }

    wrap = mount(
      <MemoryRouter >
        <Header 
          history = {history}
        />
      </MemoryRouter> 
    )

    wrap.find("WithStyles(ForwardRef(IconButton))").at(1).simulate('click');
    wrap.find("WithStyles(ForwardRef(Menu))").at(0).find("SignedOut").find("WithStyles(ForwardRef(MenuItem))").simulate('click');
    expect(history.push.mock.calls[0][0]).toEqual('/signin');
  });

})


describe('Header logged in', () => {
  const authUser = {
    cfId: "1fbyawFlFR0YdMYPZbtG",
    status: "current",
    iss: "https://securetoken.google.com/givingtree-cfs",
    aud: "givingtree-cfs",
    auth_time: 1585185093,
    user_id: "QJbnn1tU1EUCnllEGBAW4ZuLMG42",
    sub: "QJbnn1tU1EUCnllEGBAW4ZuLMG42",
    iat: 1585437646,
    exp: 1585441246,
    email: "sprechernd1@gcc.edu",
    email_verified: false,
    firebase: {
      identities: {
        email: ["sprechernd1@gcc.edu"]
      },
      sign_in_provider: "password"
    }
  }

  it('goes to Profile from menu', () => { 
    const history = {
      push : jest.fn()
    }

    let wrap = mount(
      <MemoryRouter >
        <Header 
          history = {history}
          authUser={authUser}
        />
      </MemoryRouter> 
    )
    wrap.find("WithStyles(ForwardRef(IconButton))").at(1).simulate('click');
    expect(wrap.find("WithStyles(ForwardRef(Menu))").at(0).props().anchorEl).not.toBeFalsy();
    wrap.find("WithStyles(ForwardRef(Menu))").at(0).find("SignedIn").find("WithStyles(ForwardRef(MenuItem))").at(0).simulate('click');
    expect(wrap.find("WithStyles(ForwardRef(Menu))").at(0).props().anchorEl).toBeFalsy();
  });

  it('Logs out from menu', () => { 
    const history = {
      push : jest.fn()
    }

    let wrap = mount(
      <MemoryRouter >
        <Header 
          history = {history}
          authUser={authUser}
        />
      </MemoryRouter> 
    )
    wrap.find("WithStyles(ForwardRef(IconButton))").at(1).simulate('click');
    wrap.find("WithStyles(ForwardRef(Menu))").at(0).find("SignedIn").find("WithStyles(ForwardRef(MenuItem))").at(1).simulate('click');
    expect(wrap.find("WithStyles(ForwardRef(Menu))").at(0).props().anchorEl).toBeFalsy();
  });

});

