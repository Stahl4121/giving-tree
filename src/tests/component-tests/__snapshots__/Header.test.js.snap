// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header logged out Renders correctly 1`] = `ReactWrapper {}`;
